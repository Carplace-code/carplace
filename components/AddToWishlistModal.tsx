"use client";

import { Version } from "@prisma/client";
import { Heart } from "lucide-react";
import { useState } from "react";
import { toast } from "sonner";

import { Button } from "@/components/ui/button";
import { Dialog, DialogContent, DialogDescription, DialogHeader, DialogTitle } from "@/components/ui/dialog";
import { useAddToWishlist, useCreateWishlist, useGetWishlists, useRemoveFromWishlist } from "@/hooks/useWishlists";

import { Input } from "./ui/input";

interface AddToWishlistModalProps {
  versionId: Version["id"];
}

export default function AddToWishlistModal({ versionId }: AddToWishlistModalProps) {
  const [open, setOpen] = useState(false);
  const [newWishlistName, setNewWishlistName] = useState("");
  const [newWishlistDescription, setNewWishlistDescription] = useState("");

  const { data: userWishlists, isLoading: loadingUserWishlists } = useGetWishlists({
    include: {
      items: true,
    },
  });
  const { mutate: createWishlist } = useCreateWishlist();
  const { mutate: addToWishlist } = useAddToWishlist();
  const { mutate: removeFromWishlist } = useRemoveFromWishlist();

  const isInWishlist =
    !loadingUserWishlists &&
    (userWishlists || []).some((wishlist) => wishlist.items?.map((item) => item.versionId).includes(versionId));

  const onButtonClick = () => {
    if (loadingUserWishlists) return; // Prevent opening dialog while loading

    if (!isInWishlist) {
      setOpen(true);
      return;
    }

    removeFromWishlist(
      {
        wishlistId:
          userWishlists?.find((wishlist) => wishlist.items?.some((item) => item.versionId === versionId))?.id || "",
        versionId,
      },
      {
        onSuccess: () => {
          toast.success("Eliminado de la lista de deseos");
        },
        onError: () => {
          toast.error("Error al eliminar de la lista de deseos");
        },
      },
    );
  };

  return (
    <Dialog open={open} onOpenChange={setOpen}>
      <Button variant="ghost" size="default" className="ml-2 h-auto p-1" onClick={onButtonClick}>
        <Heart className={`h-4 w-4 ${isInWishlist ? "fill-current text-red-500" : ""}`} />
      </Button>
      <DialogContent className="sm:max-w-md">
        <DialogHeader>
          <DialogTitle>Añadir a una lista</DialogTitle>
          <DialogDescription>Elige una lista a la que añadir el auto o crea una nueva!</DialogDescription>
        </DialogHeader>

        <div className="space-y-4">
          {loadingUserWishlists ? (
            <div className="py-4 text-center">Cargando listas de deseos...</div>
          ) : (
            <>
              {userWishlists?.map((wishlist) => (
                <div
                  className="flex items-center justify-between rounded-lg border px-3 py-2 hover:bg-gray-50"
                  key={wishlist.id}
                >
                  <div className="flex flex-col">
                    <span>{wishlist.name}</span>
                    <span className="text-sm text-gray-500">{wishlist.description}</span>
                  </div>
                  <Button
                    onClick={() =>
                      addToWishlist(
                        { wishlistId: wishlist.id, versionId },
                        {
                          onSuccess: () => {
                            setOpen(false);
                            toast.success(`Añadido a ${wishlist.name}`);
                          },
                          onError: () => {
                            toast.error("Error al añadir a la lista");
                          },
                        },
                      )
                    }
                  >
                    Añadir
                  </Button>
                </div>
              ))}

              {/* Horizontall divider */}
              <div className="my-4 border-t" />

              {/* Create new wishlist form */}
              <div className="flex w-full items-center gap-2">
                <Input
                  placeholder="Nombre"
                  value={newWishlistName}
                  onChange={(e) => setNewWishlistName(e.target.value)}
                  className="flex-grow"
                />
                <Button
                  onClick={() => {
                    createWishlist(
                      { name: newWishlistName, description: newWishlistDescription },
                      {
                        onSuccess: async (newWishlist) => {
                          addToWishlist(
                            { wishlistId: newWishlist.id, versionId },
                            {
                              onSuccess: () => {
                                setOpen(false);
                                setNewWishlistName("");
                                setNewWishlistDescription("");
                                toast.success(`Lista creada y añadido a ${newWishlist.name}`);
                              },
                              onError: () => {
                                toast.error("Error al añadir a la nueva lista");
                              },
                            },
                          );
                        },
                        onError: () => {
                          toast.error("Error al crear la lista");
                        },
                      },
                    );
                  }}
                  className="text-sm"
                  disabled={loadingUserWishlists || !newWishlistName.trim() || !newWishlistDescription.trim()}
                >
                  Crear y añadir
                </Button>
              </div>
              <Input
                placeholder="Descripción"
                value={newWishlistDescription}
                onChange={(e) => setNewWishlistDescription(e.target.value)}
                className="mt-2 w-full"
              />
            </>
          )}
        </div>
      </DialogContent>
    </Dialog>
  );
}
