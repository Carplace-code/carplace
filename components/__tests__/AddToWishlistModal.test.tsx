import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render } from "@testing-library/react";
import { describe, it, vi } from "vitest";

import AddToWishlistModal from "../AddToWishlistModal";

// Mock hooks
vi.mock("@/hooks/useWishlists", () => ({
  useGetWishlists: () => ({
    data: [],
    isLoading: false,
  }),
  useCreateWishlist: () => ({ mutate: vi.fn() }),
  useAddToWishlist: () => ({ mutate: vi.fn() }),
  useRemoveFromWishlist: () => ({ mutate: vi.fn() }),
}));

// Setup wrapper
const queryClient = new QueryClient();
const wrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);

describe("AddToWishlistModal", () => {
  it("renders correctly and matches snapshot", () => {
    const { container } = render(<AddToWishlistModal versionId="v123" />, { wrapper });
    expect(container).toMatchSnapshot();
  });
});
