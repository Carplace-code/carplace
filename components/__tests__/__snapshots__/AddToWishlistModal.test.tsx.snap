// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddToWishlistModal > renders correctly and matches snapshot 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-smfont-medium transition-all disabled:pointer-events-none disabled:opacity-50[&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0[&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px]aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 has-[>svg]:px-3 ml-2 h-auto p-1"
    data-slot="button"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-heart h-4 w-4"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
      />
    </svg>
  </button>
</div>
`;
